main.py

```diff
@@ -808,6 +784,8 @@ def handle_interface_link_up(self, interface):
         Handler for interface link_up events
         """
         for evc in self.get_evcs_by_svc_level():
+            if evc.archived:
+                continue
```

Comment:
Even though the archived EVCs are being deleted from `self.circuits`, I think we can't get rid of statements which check for `archive` value. [self.get_evcs_by_svc_level()](https://github.com/kytos-ng/mef_eline/blob/820d024b3a1c861fbd4ad9fbbe2cb31301b31078/main.py#L76) can still give a view of the deleted EVCs and some methods are going to encounter archived EVCs.

Reply:
> self.get_evcs_by_svc_level() can still give a view of the deleted EVCs and some methods are going to encounter archived EVCs.

It should no longer have deleted EVCs (after your refactoring) since it's creating the sorted list based on `self.circuits`. This method here was for sorting EVCs based on their service_level importance, and when handling link up|down only the existing non-archived enabled EVCs are supposed to be processed. Speaking of which can you refactor `self.get_evcs_by_svc_level()` to only include enabled EVCs? No need to parametrize any arg, all call sites usage here they're all expecting enabled EVCs when getting this list (if one day that changes then we can consider including extra args)

Can you double check the case where you still ended up seeing an archived EVC here?

Reply:
> Is sorted() thread safe? I saw that it is, if the keys are python defined classes because of GIL but not sure.

In `cpython` the implementation creates a shallow list copy before sorting in place, so it shouldn't be an issue here, and since this method is just dealing with whatever is still in memory it's OK, but the underlying `evc` instance indeed can have other threads trying to write, so indeed there's a potential race condition if an EVC were to get concurrently deleted and an UNI event link up or down is still being processed it might end up with it deleted but with an incoherent `active` state in the DB. 

So yes, I believe it might be worth also including a `with evc.lock:` when handling `self.handle_interface_link_up(iface)` and `self.handle_interface_link_down(iface)`. Good observation.

Reply:
Got it. I could not hit the archived EVC being analyzed now.